### ISSUE TITLE
[BUG] Exception running mlflow.pyfunc.load_model with custom Python model

### ISSUE THREAD

**Issue by PJPRoche** [mlflow/mlflow#6947]
### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

1.24.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux 5.4.0-1090-azure / Windows 10
- **Python version**: 3.8.10

The pip requirements inferred by mlflow are:
mlflow
astunparse==1.6.3
cloudpickle==2.2.0
dill==0.3.2
ipython==7.22.0
pyspark==3.2.1
sdv==0.17.0

### Describe the problem

I am using MLFlow to train a GAN neural network model and logging the experiment with all its artifacts, which includes the pickled custom GAN model class. I am using MLFlow in Azure Databricks for all the training.

After training, I am downloading all the artifacts so I can run some local inference. I am creating a conda environment using the inferred pip requirements (see the list in the above "System information" section) that gets bundled into the trained logged model artifacts. I then run the mlflow.pyfunc.load_model function and try get some predictions from the model.

I appreciate there are a number of factors here: platforms / environments / custom model classes which makes this hard to isolate and maybe for other to reproduce issue. However, I thought I would share in case others are having the same issue, or if someone has an idea how I can get to the bottom of the issue!! I have tried and debugged a lot of things but am not having any success in solving the problem.

Thanks for any help/suggestions

### Code to reproduce issue

This is the Databricks notebook I am using to train the model
```
# Databricks notebook source
pip install sdv==0.17.0

# COMMAND ----------

# DBTITLE 1,Install and Load Libraries
import logging
import mlflow
import mlflow.pyfunc

from mlflow.tracking.client import MlflowClient

from sdv.tabular import CTGAN
from sdv.demo import load_tabular_demo

from sys import version_info

# COMMAND ----------

# DBTITLE 1,Configure logging
# setting logging to DEBUG level whilst in development
logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)

# turn off/set to error other modules
logger = logging.getLogger("py4j").setLevel(logging.ERROR)
logger = logging.getLogger("rdt").setLevel(logging.ERROR)
logger = logging.getLogger("urllib3v").setLevel(logging.ERROR)

# COMMAND ----------

mlflow_experiment_name = "/Shared/SyntheticGAN"

SDV_MODEL_NAME = "sdv_model.pkl"

# name of sub-folder to save the MLFlow model under the experiment id
MLFLOW_PYFUNC_MODEL_PATH = "sdv_mlflow_pyfunc"

# Parameter grid for Neural Network
PARAM_GRID = dict(
  epochs = 1
)

# COMMAND ----------

PYTHON_VERSION = "{major}.{minor}.{micro}".format(major=version_info.major,
                                                  minor=version_info.minor,
                                                  micro=version_info.micro)

mlflow.set_experiment(mlflow_experiment_name)

# COMMAND ----------

# Create an `artifacts` dictionary that assigns a unique name to the saved SDV model file.
# This dictionary will be passed to `mlflow.pyfunc.log_model`, which will copy the model file
# into the new MLflow experiment model's directory.
artifacts = {
    "sdv_model": SDV_MODEL_NAME
}

# COMMAND ----------

# Define the model class
# https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#
# https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#pyfunc-create-custom
class SDVWrapper(mlflow.pyfunc.PythonModel):
    """Defines the custom wrapper class."""
    def load_context(self, context):
        from sdv.tabular import CTGAN
        self.sdv_model = CTGAN.load(context.artifacts["sdv_model"])

    def predict(self, context, model_input):
        return self.sdv_model.sample(num_rows=model_input)

# COMMAND ----------

data = load_tabular_demo('student_placements')

# COMMAND ----------

# Start a new MLflow training run 
with mlflow.start_run() as run:
  
    run_id = "runs:/" + run.info.run_id + "/sdv_mlflow_pyfunc"
    
    model = CTGAN()
    model.fit(data)
    model.save(SDV_MODEL_NAME)
    
    # Log the MLflow Model
    # https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#mlflow.pyfunc.log_model
    mlflow.pyfunc.log_model(
        artifact_path=MLFLOW_PYFUNC_MODEL_PATH, 
        python_model=SDVWrapper(), 
        artifacts=artifacts,
    )
  

# COMMAND ----------

run_id

# COMMAND ----------

model = mlflow.pyfunc.load_model(run_id)

# COMMAND ----------

model.predict(10)

```

This is the local python script I am using to make inferences with

```
import mlflow

logged_model = 'PATH_TO_LOCALLY_SAVED_TRAINED_MODEL_ARTIFACTS_FOLDER'

loaded_model = mlflow.pyfunc.load_model(logged_model)
print(loaded_model.predict(10))
```

### Stack trace

```
Traceback (most recent call last):
  File "C:\Users\######\Miniconda3\envs\sdv-demo\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\######\Miniconda3\envs\sdv-demo\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\######\.vscode\extensions\ms-python.python-2022.2.1924087327\pythonFiles\lib\python\debugpy\__main__.py", line 45, in <module>
    cli.main()
  File "c:\Users\######\.vscode\extensions\ms-python.python-2022.2.1924087327\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 444, in main
    run()
  File "c:\Users\######\.vscode\extensions\ms-python.python-2022.2.1924087327\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 285, in run_file
    runpy.run_path(target_as_str, run_name=compat.force_str("__main__"))
  File "C:\Users\######\Miniconda3\envs\sdv-demo\lib\runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "C:\Users\######\Miniconda3\envs\sdv-demo\lib\runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "C:\Users\######\Miniconda3\envs\sdv-demo\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\######\repos\###########/run_models/sample_data_generator/run_mlflow_test.py", line 5, in <module>
    loaded_model = mlflow.pyfunc.load_model(logged_model)
  File "C:\Users\######\Miniconda3\envs\sdv-demo\lib\site-packages\mlflow\pyfunc\__init__.py", line 710, in load_model
    model_impl = importlib.import_module(conf[MAIN])._load_pyfunc(data_path)
  File "C:\Users\######\Miniconda3\envs\sdv-demo\lib\site-packages\mlflow\pyfunc\model.py", line 280, in _load_pyfunc
    python_model.load_context(context=context)
  File "<command-1315971304203110>", line 8, in load_context
  File "C:\Users\######\Miniconda3\envs\sdv-demo\lib\site-packages\sdv\tabular\base.py", line 891, in load
    model = cloudpickle.load(f)
TypeError: _restore() takes 3 positional arguments but 4 were given
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [X] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

**Comment by dbczumar** [comment]
Hi @PJPRoche, thank you for the very detailed bug report! To further isolate the problem, I have a few questions / things to try:

1. Are you running the same version of Python in your local conda environment that you're running on your Databricks cluster?
2. Can you verify that the version of cloudpickle in your local conda environment is identical to the version running on your Databricks cluster?
3. To remove MLflow from the equation for testing purposes, can you try copying the saved `CTGAN` model written via `model.save(SDV_MODEL_NAME)` from Databricks to your local filesystem and try to load it within your conda environment and see if it succeeds?

**Comment by PJPRoche** [comment]
Hi @dbczumar, in answer to your questions:

1. yes, it is the exact same 3.8.10 version both on Databricks and in my local conda environment
2. yes, it is 2.2.0 both locally and in Databricks. I have created the local environment with the requirements.txt produced by the MLFLow run. I have also created a more complete environment with all the Python modules from the Databricks environment, which is a lot more as Databricks pre-installs a lot of common modules... Neither of these seem to work, and I get the same issue when I try load the .pkl locally.
3. I have done this too and it does **not** work. I can only really conclude that it is not MLFlow per se but something with the recreation of the environment locally, but I can't think of what more now to test!!

I also did one more test (and reversed the workflow) and ran MLFlow locally and exported the trained model to the remote Databricks, loaded the .pkl and was able to make inferences OK. 

So, in summary, running everything on Databricks (training and inference) works fine. Running everything locally (training and inference) works fine. Training locally and exported to Databricks works fine. The issue seems only one way, copying the trained model/.pkl from Databricks to the local computer and running it. 

Unfortunately, our required workflow means we need to train the model remotely and then save down the trained model to use for inferences. I appreciate that this now does not seem like MLFlow related in itself but any suggestions welcome - it is all quite odd!!

**Comment by dbczumar** [comment]
Hi @PJPRoche, thank you for isolating this issue. I recommend filing a ticket on https://github.com/sdv-dev/SDV/issues, as this seems to be a portability problem impacting the `sdv` library. Because your Python, sdv, and cloudpickle versions are consistent across Databricks and your local environment, I'm unfortunately unable to advise further. Thank you for using MLflow, and I hope that the sdv maintainers can be of more help for this particular issue! :)
